### ISSUE TITLE
issue parsing output map variables with dotted notation in the key

### ISSUE THREAD

**Issue by bcenker** [hashicorp/terraform#10876]
running terraform 0.7.13

We have a example case wherein we would like to have a map of FQDNs to route53 zone ids.  We would like this map to be consumed either as a module or from a remote state file.  We have found that when we generate a map with dotted notation in the key, we are able to use that map variable locally, and declare it as an output, but when we attempt to consume it either as a module or from a remote state file, it has been parsed as a nested map.

Example
```hcl
resource "null_resource" "null" { }

variable "zone_id" {
  type = "map"
  default = {
    "use1.amz.blah.com" = "111111111"
    "use2.amz.blah.com" = "222222222"
    "usw2.amz.blah.com" = "333333333"
    "anotherdomain.com" = "444444444"
  }
}

output "zone_id" {
  value = "${var.zone_id}"
}

output "zone_id_use1.amz.blah.com" {
  value = "${var.zone_id["use1.amz.blah.com"]}"
}
```
This generates the following output as expected:
```
Outputs:

zone_id = {
  anotherdomain.com = 444444444
  use1.amz.blah.com = 111111111
  use2.amz.blah.com = 222222222
  usw2.amz.blah.com = 333333333
}
zone_id_use1.amz.blah.com = 111111111
```

However, when trying to consume this variable from the remote state file:
```hcl
provider "null" "null" {}

data "terraform_remote_state" "route53" {
  backend = "local"
  config = {
    path = "../route53/terraform.tfstate"
  }
}

output "map_as_passed_in_by_remote_state" {
  value = "${data.terraform_remote_state.route53.zone_id}"
}

output "single_value_from_map_read_from_remote_state" {
  value = "${data.terraform_remote_state.route53.zone_id_use1.amz.blah.com}"
}
```
we receive in the following:
```
Outputs:

map_as_passed_in_by_remote_state = {
  anotherdomain = map[com:444444444]
  use1 = map[amz:map[blah:map[com:111111111]]]
  use2 = map[amz:map[blah:map[com:222222222]]]
  usw2 = map[amz:map[blah:map[com:333333333]]]
}
single_value_from_map_read_from_remote_state = 111111111
```

As you can see from the remote tfstate file, the output is represented properly:
```
            "outputs": {
                "zone_id": {
                    "sensitive": false,
                    "type": "map",
                    "value": {
                        "anotherdomain.com": "444444444",
                        "use1.amz.blah.com": "111111111",
                        "use2.amz.blah.com": "222222222",
                        "usw2.amz.blah.com": "333333333"
                    }
                },
```

but this it what appears to be getting parsed into (from the consumer terraform.tfstate):
```
            "outputs": {
                "map_as_passed_in_by_remote_state": {
                    "sensitive": false,
                    "type": "map",
                    "value": {
                        "anotherdomain": {
                            "com": "444444444"
                        },
                        "use1": {
                            "amz": {
                                "blah": {
                                    "com": "111111111"
                                }
                            }
                        },
                        "use2": {
                            "amz": {
                                "blah": {
                                    "com": "222222222"
                                }
                            }
                        },
                        "usw2": {
                            "amz": {
                                "blah": {
                                    "com": "333333333"
                                }
                            }
                        }
                    }
                },
```

We have only seen this occur when using dotted notation in the key.  Replacing dots with dashes will result in the proper passing and parsing of the map variable, but it's not ideal to use dashes when the goal is to use the FQDN as the key reference.  Is the behavior we are seeing expected?  We did not see a limitation in the documentation that you could not use dotted notation in map keys, and as mentioned, we are able to use this map variable normally within the terraform it is declared in. 

Thanks!

**Comment by jaymecd** [comment]
Same problem on `0.11` version.
Any progress on this issue?
Or advice how to deal with dotted keys would be appreciated.

**Comment by zytek** [comment]
Re https://github.com/terraform-providers/terraform-provider-aws/issues/5918

This issue makes it impossible to pass maps from data source to resource, apart from other cases mentioned here. 

Could some experienced dev propose solution/workaround or provide more contexts around this? I could participate in testing/fixing with some guidance.

**Comment by Cib0rg** [comment]
Same problem. In two years (!) this issue wasn't even assigned to anybody? I found fix for resource, it merged to master. But I cannot found code block similiar to that fix =(

**Comment by zytek** [comment]
@Cib0rg can you point for the resource fix merged to master?

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

There are a few different things going on with this issue:

One problem here is that outputs with non-identifier names (such as including periods) should always have been invalid. There was a missing validation rule in prior versions of Terraform that caused them to be allowed, but that then caused incorrect downstream behavior as we see here. This particular detail is "fixed" in master by adding the correct validation rule that would consider the output declared in this example to be invalid:

```
Error: Invalid output name

  on output-dots.tf line 17, in output "zone_id_use1.amz.blah.com":
  17: output "zone_id_use1.amz.blah.com" {

A name must start with a letter and may contain only letters, digits,
underscores, and dashes.
```

However, even with that corrected validation in place there remained an issue when returning a map whose keys themselves contained dots, like this:

```hcl
variable "zone_id" {
  type = "map"
  default = {
    "use1.amz.blah.com" = "111111111"
    "use2.amz.blah.com" = "222222222"
    "usw2.amz.blah.com" = "333333333"
    "anotherdomain.com" = "444444444"
  }
}

output "zone_ids" {
  value = "${var.zone_id}"
}
```

This would propagate through as expected into the state, but then be misinterpreted at the other end by the `terraform_remote_state` data source, since the Terraform state format in Terraform v0.11 and prior uses dot-delimited strings as attribute path keys, and so it would decode them incorrectly.

That second issue is now fixed in master ready for the forthcoming v0.12.0 release. I verified it with the v0.12.0-alpha2 prerelease build using this configuration against a state file created by the configuration I showed above:

```hcl
data "terraform_remote_state" "example" {
  backend = "local"
  config = {
    path = "foo.tfstate"
  }
}

output "zone_ids" {
  value = data.terraform_remote_state.example.outputs.zone_ids
}
```

```
$ terraform apply
data.terraform_remote_state.example: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

zone_ids = {
  "anotherdomain.com" = "444444444"
  "use1.amz.blah.com" = "111111111"
  "use2.amz.blah.com" = "222222222"
  "usw2.amz.blah.com" = "333333333"
}
```

The map now emerges the other end with the same keys it has at input, because the new internal representation of attributes no longer uses the dot-separated keys I mentioned before.

A necessary consequence of this fix was to move the output keys into a map attribute called `outputs` in the data source, rather than just just exposing them directly as weird extra dynamic attributes. This allowed both this fix and fixes for several other odd behaviors such as conflicts when exporting outputs with names like `backend` and `config` which are also arguments to the `terraform_remote_state` data source. A configuration upgrade tool will be included in the final v0.12.0 release which will be able to detect and update remote state references to include that extra `.outputs` attribute selector.

Since this is fixed in the master branch, I'm going to close this out. Thanks for reporting this, and sorry for the delay while we got all the layers in between ready to deal with this.
